### ISSUE TITLE
RNN Initial State

### ISSUE THREAD

**Issue by vegapit** [LaurentMazare/tch-rs#87]
Hello,

I am looking at implementations of the **seq2seq** model and if I am not mistaken, a functionality seems to be missing in the current `RNN` trait for that purpose: The ability to set an initial state before performing a `seq` update. Here is a method for `GRU` (strongly inspired by your code) that could do just that:

```rust
impl GRU {

    pub fn seq_init(&self, input: &Tensor, in_state: &GRUState) -> (Tensor, GRUState) {
        let GRUState(h) = in_state;
        let batch_dim = input.size()[0];
        let shape = [1, batch_dim, self.hidden_dim];
        let (output, h) = input.gru(
            &h,
            &[&self.w_ih, &self.w_hh, &self.b_ih, &self.b_hh],
            self.config.has_biases,
            self.config.num_layers,
            self.config.dropout,
            self.config.train,
            self.config.bidirectional,
            self.config.batch_first,
        );
        (output, GRUState(h))
    }

}
```
I am still running tests on it, so not sure yet if it is the best solution...

**Comment by LaurentMazare** [comment]
Indeed, I think that for most of the use case the initial state is zero when using `seq` and `step` can be used for inference but your `seq_init` may be handy too.
I just pushed some changes to add `seq_init` and derive the implementation of `seq` from `seq_init` and `zero_state`, these changes are fairly similar to your suggestion.
Thanks for reporting this!

**Comment by vegapit** [comment]
No problem, thank you for pushing the changes. I will let you know if I encounter any other demanding cases.

**Comment by vegapit** [comment]
I am encountering a very strange problem when trying to initialise `GRUState` from outside the `tch-rs` project. The initialisation of the *Struct Tuple* fails with the following error:
```
error[E0423]: expected function, found struct `GRUState`
let mut state = GRUState( Tensor::zeros(&[1,BATCH_SIZE,EMBEDDING_DIMENSION], (Kind::Float, Device::Cpu)) );
   |                         ^^^^^^^^ did you mean `GRUState { /* fields */ }`?
```
I rarely use *Struct Tuples* so I am not quite sure what the problem could be.

**Comment by LaurentMazare** [comment]
The error message is very confusing but I think this is caused by the inner part of `GRUState` not being public, e.g. if you change the definition of `GRUState` to be something like the following the error will probably go away:
```rust
pub struct GRUState(pub Tensor);
```

If you have some worthy use case for creating a state with some given tensor we can certainly expose this.

**Comment by vegapit** [comment]
That was the issue, great stuff.

The use case is for bootstrapping the result of a GRU unit to compute its next n time steps: `(x_{n},h_{n}) = GRU o GRU o ... o GRU(x_0,h_0)`. This is to implement the decoder part of a seq2seq model.

If my understanding is correct, I can do this by looping `seq_init` performing one time step at a time, but I am not getting the output dimensions I am expecting. Assuming that the GRU has 1 layer, is unidirectional and input_size=hidden_size=dim, The GRU doc for [pytorch](https://pytorch.org/docs/stable/nn.html?highlight=gru#torch.nn.GRU) shows that for an input shape of (seq_len, batch, dim) and state space of (1, batch, dim), we can expect an output shape (seq_len, batch, dim). When I use the `seq_init` function with these input and state shapes, I am getting (batch,batch,dim) which is a little peculiar.

 Also, according to the same doc, the shape of the state tensor varies according to the chosen GRU config (number of layers and number of directions), but I am not seeing this coming through in the `seq` code. Do you manage it in the `Tensor.gru` method?

**Comment by LaurentMazare** [comment]
I think that by default the python api uses `batch_first=False` whereas we set it to true in this crate. This probably explains the weird shapes you're seeing.

**Comment by vegapit** [comment]
That was exactly the problem, thanks for the clarification. It is working as expected now.

**Comment by vegapit** [comment]
I confirm that the `GRU` Struct does not work with `bidirectional=true`. There is a thread on the [pytorch C++ API](https://github.com/pytorch/pytorch/issues/17998) project regarding this specific error, but I can not make it work even by downloading the nightly versions of libtorch.
